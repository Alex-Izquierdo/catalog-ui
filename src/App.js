import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { Routes } from './Routes';
import './App.scss';
import PortalNav from './SmartComponents/ServicePortal/PortalNav';
import { Main } from '@red-hat-insights/insights-frontend-components';
import { NotificationsPortal } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import { Grid, GridItem } from '@patternfly/react-core';
import { MIN_SCREEN_HEIGHT } from './constants/ui-constants';
import '@red-hat-insights/insights-frontend-components/components/Notifications.css';

class App extends Component {

  componentDidMount () {
    insights.chrome.init();
    insights.chrome.identifyApp('service-portal');
  }

  componentWillUnmount () {
    this.appNav();
    this.buildNav();
  }

  render () {
    return (
      <React.Fragment>
        <NotificationsPortal />
        <Main style={ { marginLeft: 0, padding: 0 } }>
          <Grid style={ { minHeight: MIN_SCREEN_HEIGHT } }>
            <GridItem style={ { backgroundColor: '#FFFFFF' } } sm={ 4 } md={ 4 } lg={ 2 } xl={ 2 }>
              <PortalNav />
            </GridItem >
            <GridItem sm={ 8 } md={ 8 } lg={ 10 } xl={ 10 }>
              <Routes childProps={ this.props } />
            </GridItem>
          </Grid>
        </Main>
      </React.Fragment>
    );
  }
}

App.propTypes = {
  history: PropTypes.object
};

/**
 * withRouter: https://reacttraining.com/react-router/web/api/withRouter
 * connect: https://github.com/reactjs/react-redux/blob/master/docs/api.md
 *          https://reactjs.org/docs/higher-order-components.html
 */
export default withRouter (connect()(App));
