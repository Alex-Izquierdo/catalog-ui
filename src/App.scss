@import '~@redhat-cloud-services/frontend-components/index.css';

.elipsis-text-overflow {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.card-link, .card-link:hover, .card-link:active {
  text-decoration: none !important;
  cursor: pointer !important;
  color: #282d33 !important;
  font-weight: 400 !important;
}

.pf-c-dropdown__menu {
  a {
    text-decoration: none !important;
  }
}

.text-align-right {
  text-align: right;
}

.modal-form-right-align {
  .pf-m-action:last-child {
    .data-driven-forms__pf4-button-group:first-child {
      margin-left: auto
    }
  }
}

.line-clamp {
  display: -webkit-box;
  -webkit-line-clamp: 6;
  -webkit-box-orient: vertical;  
  overflow: hidden;
  word-wrap: break-word;
}

.content-gallery {
  .content-gallery-card {
    word-break: break-all;
    height: 330px;
    .card-link {
      height: 100%;
    }
  }
}
.disabled-color {
  color: #d2d2d2;
}

.destructive-color {
  &:not(:disabled) {
    color: #FF0000 !important;
  }
}


.card-image {
  height: 40px;
  &.hide {
    display: none;
    float: left;
    height: 0px;
    width: 0px;
  }
}

.data-list-expand-fix {
  .pf-c-data-list__toggle {
    align-self: flex-start;
    margin-top: 0;
  }
  .pf-c-data-list__expandable-content-body {
    padding-left: 0;
    padding-right: 0;
    padding-top: 16px;
    padding-bottom: 16px;
  }
}

.cell-grow {
  flex-grow: 20 !important;
  .pf-l-split {
    .pf-l-split__item:nth-child(2) {
      width: 100%;
    }
  }
}

.requests-table {
  width: 100%;
  th.action-row {
    width: 1px;
  }
  .requests-step-label {
    span:first-child {
      width: 24px;
      border-radius: 50%;
      height: 24px;
      display: inline-block;
      vertical-align: middle;
      text-align: center;
      background-color: #EDEDED;
      margin-right: 4px;
      line-height: 26px;
    }
  }
  tr {
    td:first-child, td:last-child, th:first-child, th:last-child {
      padding-left: calc(32px + 64px);
      padding-right: 32px;
    }
  }
  thead {
    th {
      color: #292E34;
      font-family: Overpass;
      font-size: 14px;
      line-height: 22px;
      font-weight: normal;
      padding-bottom: 16px;
    }
    border-bottom: 1px solid #EDEDED;
  }
  tbody {
    tr {
      font-size: 14px;
      &.finished {
          .requests-step-label {
            span:first-child {
              background-color: var(--pf-global--primary-color--100);
              color: #FFFFFF
            }
          }
          &.failed {
            .requests-step-label {
              span:first-child {
                background-color: var(--pf-global--danger-color--100);
              }
            }
        }
      }
      td {
        padding-top: 8px;
        padding-bottom: 8px;
      }
      &:first-child {
        td {
          padding-top: 16px;
        }
      }
    }
  }
}

.toolbar-filter-input-group {
  position: relative;
  width: 300px;
  height: 36px;
  input {
    position: absolute;
    width: 300px;
    background: transparent;
  }
  .filter-icons-container {
    position: absolute;
    right: 8px;
    top: 8px;
    svg {
      z-index: 1000;
      fill: #72767B;
    }
    &:hover {
      svg.clear-filter {
        fill: var(--pf-global--Color--100);
        cursor: pointer;
      }
    }
  }
}

.toolbar-override {
  display: flex;
  align-items: center;
  > div, > a {
    &:not(:last-child) {
      margin-right: 16px;
    }
  }
}

.button-with-spinner {
  display: flex;
  align-items: center;
  .ins-c-spinner::before {
    height: 16px;
    width: 16px;
    margin-left: 8px;
    border-color: #ededed;
    border-top-color: var(--pf-c-button--disabled--BackgroundColor);
  }
}

.portfolio-item-progress {
  position: relative;
  height: 330px;
  .content-gallery-card {
    position: absolute;
    width: 100%;
    &.progress-overlay {
      background-color: rgba(210, 210, 210, .5);
      z-index: 1;
      cursor: progress;
    }
  }
}

.disabled-link {
  pointer-events: none
}

.orders-list {
  background-color: var(--pf-global--BackgroundColor--100)
}

/**
* Update DDF select styles for select component
*/
.filter-select {
  width: 300px;
  .filter-value-container {
    position: absolute;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: flex-start;
  }
  .ddorg__pf4-component-mapper__select__indicators {
    margin-left: auto
  }
  .ddorg__pf4-component-mapper__select__placeholder {
    margin-left: 29px !important;
  }
  .ddorg__pf4-component-mapper__select__single-value {
    margin-left: 22px !important;
  }
  .ddorg__pf4-component-mapper__select__control::before {
    border-right: none !important;
  }
  .ddorg__pf4-component-mapper__select__control {
    min-height: 36px !important;
  }
}

//pf-4 fixes
/**
* non working pf display modifier: https://www.patternfly.org/v4/documentation/core/utilities/display#display-block
*/
.pf-u-display-block {
  display: block
}
// Image uploader
.image-upload-button {
  border: none;
  position: absolute;
  top: 0;
  left: 0;
  padding-left: 0;
  width: 100%;
  height: 100%;
  background-color: transparent;
  svg {
    background-color: rgba(255,255,255,0.8);
    position: absolute;
    top: 0;
    left: 0;
  }
}

.image-preview {
  height: 64px;
  max-width: 300px;
  object-fit: cover;
}

.image-upload-spinner:before{
  position: absolute;
  top:0;
  left: 0;
  width: 16px;
  height: 16px;
}

.share-group-modal{
  .height-normalize {
    height: 38px;
  }
  .share-column {
    flex-grow: 1;
    &:not(:first-child) {
      max-width: 200px;
      margin-left: 8px;
    }
    h4 {
      margin: 0
    }
  }
  &.group {
    border-top: 1px solid #d2d2d2;
    padding-top: 24px;
  }
}

.app-tabs {
  border-bottom: 1px solid lightgrey;
}

.empty-state {
  height: calc(100% - 197px)
}

.orders {
  &.bg-fill {
    background-color: var(--pf-global--Color--light-100);
  }
  &.separator {
    border-bottom: 2px solid var(--pf-global--Color--light-200);
  }
}

/**
* Side nav styles
*/
.orders-side-nav-category,
.orders-side-nav-link {
  color: #151515;
  font-weight: 500;
  line-height: 27px;
}

.orders-side-nav-category {
  font-size: 16px;
}

.orders-side-nav-category > .pf-c-nav__link {
  border-left: 4px solid transparent;
  padding: 8px 16px;
}

.orders-side-nav-category > .pf-c-nav__link,
.orders-side-nav-category.pf-m-current > .pf-c-nav__link {
  color: #151515;
  font-weight: 400;
}

.orders-side-nav-list .orders-side-nav-category > .pf-c-nav__link::after,
.orders-side-nav-list .orders-side-nav-category.pf-m-current > .pf-c-nav__link::after {
  content: none;
}

.orders-side-nav-category > .pf-c-nav__link:hover,
.orders-side-nav-item .orders-side-nav-link.pf-m-active,
.orders-side-nav-item .orders-side-nav-link:hover {
  background-color: #f8f8f8;
  border-left-color: #06c;
  color: #151515;
  font-weight: 400;
}

.orders-side-nav-item .orders-side-nav-link::after {
  background-color: transparent !important;
}

.orders-nav-context-switcher-dropdown .pf-c-dropdown__toggle {
  width: 100%;
}

.orders-nav-context-switcher > label,
.orders-nav-context-switcher .pf-c-dropdown__toggle-text,
.orders-nav-context-switcher .pf-c-dropdown__menu-item {
  color: var(--pf-global--active-color--100);
  font-weight: var(--pf-global--FontWeight--bold);
}

.orders-nav-context-switcher .pf-c-dropdown__toggle::before {
  border: 1px solid #979797;
}

.orders-nav-section-group {
  padding: 0 16px 8px 16px;
  border-left: 4px solid transparent;
}
.orders-side-nav-item.disabled {
  pointer-events: none;
  > a {
    color: var(--pf-global--Color--light-300) !important;
  }
}

@media screen and (max-width: 768px) {
  .orders-nav-layout {
    flex-direction: column;
    padding: 0 32px;
    .order-detail-nav-cotainer {
      margin-bottom: 8px;
    }
    .pf-l-split__item {
      margin-right: 0 !important;
    }
    .orders-side-nav-category > .pf-c-nav__link:hover,
    .orders-side-nav-item .orders-side-nav-link.pf-m-active,
    .orders-side-nav-item .orders-side-nav-link:hover {
      background-color: transparent;
      border-left: none;
      color: var(--pf-c-nav__tertiary-list-link--Color);
      &:after {
        position: absolute;
        bottom: 0;
        left: 0;
        display: block;
        width: 100%;
        height: var(--pf-c-nav__tertiary-list-link--after--Height);
        content: "";
        background-color: var(--pf-c-nav__tertiary-list-link--m-current--after--BackgroundColor) !important;
      }
    }
    .pf-c-nav__item.orders-nav-section-group {
      border-left: 0;
      padding-left: 0;
    }
    .pf-c-nav__item.orders-side-nav-item.orders-side-nav-category {
      float: left;
      margin-right: var(--pf-c-nav__tertiary-list-item--MarginRight);
      a {
        padding-left: 0;
        padding-right: 0;
        border-left: none;
      }
    }
  }
}

a.breadcrumb-active {
  pointer-events: none;
  color: var(--pf-c-breadcrumb__link--m-current--Color);
  text-decoration: none;
}