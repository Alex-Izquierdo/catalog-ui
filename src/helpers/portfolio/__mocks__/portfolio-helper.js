export function fetchPortfolioByName(name = '') {
  return new Promise((res) => {
    if (name === 'should-conflict') {
      return res({
        data: [{ name: 'should-conflict', id: 'conflict-id' }]
      });
    }

    return res({
      data: [],
      meta: { count: 0 }
    });
  });
}
