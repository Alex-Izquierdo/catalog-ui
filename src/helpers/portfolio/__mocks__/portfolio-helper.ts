import { ApiCollectionResponse, AnyObject } from '../../../types/common-types';

export const fetchPortfolioByName = (
  name = ''
): Promise<Partial<ApiCollectionResponse<AnyObject>>> => {
  return new Promise((res) => {
    if (name === 'should-conflict') {
      return res({
        data: [{ name: 'should-conflict', id: 'conflict-id' }]
      });
    }

    return res({
      data: [],
      meta: { count: 0, limit: 50, offset: 0 }
    });
  });
};
