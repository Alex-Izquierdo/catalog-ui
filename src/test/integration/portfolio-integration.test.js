import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';

import { testStore } from '../../utilities/store';
import { Provider } from 'react-redux';
import { mockApi } from '../../helpers/shared/__mocks__/user-login';
import {
  CATALOG_API_BASE,
  SOURCES_API_BASE,
  APPROVAL_API_BASE,
  RBAC_API_BASE
} from '../../utilities/constants';
import { MemoryRouter } from 'react-router-dom';
import App from '../../App';
import ContentGalleryEmptyState from '../../presentational-components/shared/content-gallery-empty-state';
import Portfolios from '../../smart-components/portfolio/portfolios';
import toJson from 'enzyme-to-json';
import AddPortfolioModal from '../../smart-components/portfolio/add-portfolio-modal';

import * as PortfolioHelper from '../../helpers/portfolio/portfolio-helper';
import { StyledGalleryItem } from '../../presentational-components/styled-components/styled-gallery';
import Portfolio from '../../smart-components/portfolio/portfolio';
import PortfolioEmptyState from '../../smart-components/portfolio/portfolio-empty-state';
import RemovePortfolioModal from '../../smart-components/portfolio/remove-portfolio-modal';
import EditApprovalWorkflow from '../../smart-components/common/edit-approval-workflow';

describe('Integration test for portfolio entity', () => {
  jest.useFakeTimers();
  afterEach(() => {
    //mockApi.restore();
  });
  it('should set/remove workflow to portfolio and share/unshare it', async (done) => {
    expect.assertions(16);
    const initialPortfolio = {
      id: '123',
      name: 'New portfolio',
      created_at: new Date().toString(),
      metadata: {
        user_capabilities: {
          copy: true,
          show: true,
          update: true,
          destroy: true,
          share: true,
          unshare: true,
          set_approval: true
        }
      }
    };
    jest.spyOn(global.insights.chrome, 'getUserPermissions').mockImplementation(
      () =>
        new Promise((res) =>
          res([
            {
              permission: 'catalog:portfolios:create'
            }
          ])
        )
    );
    let wrapper;
    /**
     * Mock initial open API request and souce types request
     */
    mockApi.onGet(`${CATALOG_API_BASE}/openapi.json`).replyOnce(200, {
      components: { schemas: {} }
    });
    mockApi
      .onGet(`${SOURCES_API_BASE}/source_types`)
      .replyOnce(200, { data: [] });
    /**
     * Mock initial portfolios request after user lands on portfolios page
     * Response is empty
     */
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, {
        data: [initialPortfolio],
        meta: { count: 1, limit: 50, offset: 0 }
      });
    const store = testStore();
    await act(async () => {
      wrapper = mount(
        <Provider store={store}>
          <MemoryRouter initialEntries={['/']}>
            <App />
          </MemoryRouter>
        </Provider>
      );
    });
    /**
     * One update to wait for chuk load
     */
    await act(async () => {
      wrapper.update();
    });
    /**
     * One update to update the component state
     */
    await act(async () => {
      wrapper.update();
    });
    expect(wrapper.find(Portfolios)).toHaveLength(1);
    expect(wrapper.find(StyledGalleryItem)).toHaveLength(1);

    /**
     * expand the dropdown and click on the set approval option
     */
    wrapper
      .find(`button#portfolio-${initialPortfolio.id}-toggle`)
      .simulate('click');
    /**
     * mock workflows endpoints
     * workflows for the select must be called twice because of the initial values pre-fetch
     */
    mockApi
      .onGet(
        `${APPROVAL_API_BASE}/workflows?app_name=catalog&object_type=Portfolio&object_id=${initialPortfolio.id}&filter[name][contains]=&limit=50&offset=0` // eslint-disable-line max-len
      )
      .replyOnce(200, {
        data: [],
        meta: { limit: 50, offset: 0, count: 0 }
      });
    mockApi
      .onGet(`${APPROVAL_API_BASE}/workflows?filter[name][contains]=&`)
      .reply(200, {
        data: [
          {
            id: '1',
            name: 'Workflow 1'
          },
          {
            id: '2',
            name: 'Workflow 2'
          }
        ],
        meta: {
          count: 2,
          limit: 100,
          offset: 0
        }
      });
    await act(async () => {
      wrapper
        .find('li#workflow-portfolio-action a')
        .simulate('click', { button: 0 });
    });
    jest.runAllTimers();
    await act(async () => {
      wrapper.update();
    });
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolios/edit-workflow');
    expect(wrapper.find(EditApprovalWorkflow)).toHaveLength(1);
    /**
     * open the select and choose forst option Workflow 1
     */
    await act(async () => {
      wrapper.update();
      jest.runAllTimers();
    });
    wrapper
      .find('div.ddorg__pf4-component-mapper__select__control')
      .simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
    expect(
      wrapper.find('div.ddorg__pf4-component-mapper__select__option')
    ).toHaveLength(2);

    wrapper
      .find('div.ddorg__pf4-component-mapper__select__option')
      .first()
      .simulate('click');

    /**
     * mock workflow link endpoint
     */
    mockApi.onPost(`${APPROVAL_API_BASE}/workflows/1/link`).replyOnce(200);
    await act(async () => {
      wrapper.find('form').simulate('submit');
    });
    wrapper.update();
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolios');
    expect(wrapper.find(EditApprovalWorkflow)).toHaveLength(0);
    /**
     * Open the workflow modal again and unlink the first workflow and link the second
     */

    wrapper
      .find(`button#portfolio-${initialPortfolio.id}-toggle`)
      .simulate('click');

    /**
     * mock workflows endpoints
     * workflows for the select must be called twice because of the initial values pre-fetch
     */
    mockApi
      .onGet(
        `${APPROVAL_API_BASE}/workflows?app_name=catalog&object_type=Portfolio&object_id=${initialPortfolio.id}&filter[name][contains]=&limit=50&offset=0` // eslint-disable-line max-len
      )
      .replyOnce(200, {
        data: [
          {
            id: '1',
            name: 'Workflow 1'
          }
        ],
        meta: { limit: 50, offset: 0, count: 0 }
      });
    mockApi
      .onGet(
        `${APPROVAL_API_BASE}/workflows?filter[name][contains]=&filter[id][]=1`
      )
      .replyOnce(200, {
        data: [
          {
            id: '1',
            name: 'Workflow 1'
          }
        ],
        meta: {
          count: 1,
          limit: 100,
          offset: 0
        }
      });
    await act(async () => {
      wrapper
        .find('li#workflow-portfolio-action a')
        .simulate('click', { button: 0 });
    });
    jest.runAllTimers();
    await act(async () => {
      wrapper.update();
    });
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolios/edit-workflow');
    expect(wrapper.find(EditApprovalWorkflow)).toHaveLength(1);
    /**
     * open the select and de select forst workflow and select the scond
     */
    await act(async () => {
      wrapper.update();
      jest.runAllTimers();
    });
    wrapper
      .find('div.ddorg__pf4-component-mapper__select__control')
      .simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });

    expect(
      wrapper.find('div.ddorg__pf4-component-mapper__select__option')
    ).toHaveLength(2);

    wrapper.update();
    wrapper
      .find('div.ddorg__pf4-component-mapper__select__option')
      .first()
      .simulate('click');

    wrapper
      .find('div.ddorg__pf4-component-mapper__select__option')
      .last()
      .simulate('click');

    /**
     * simulate unlink and link endpoints
     */
    mockApi.onPost(`${APPROVAL_API_BASE}/workflows/1/unlink`).replyOnce(200);
    mockApi.onPost(`${APPROVAL_API_BASE}/workflows/2/link`).replyOnce(200);
    await act(async () => {
      wrapper.find('form').simulate('submit');
    });
    wrapper.update();

    /**
     * remove existing shares in portfolio and add new one
     */

    wrapper
      .find(`button#portfolio-${initialPortfolio.id}-toggle`)
      .simulate('click');

    /**
     * mock share info endpoint
     */
    mockApi
      .onGet(`${CATALOG_API_BASE}/portfolios/${initialPortfolio.id}/share_info`)
      .replyOnce(200, [
        {
          group_name: 'Existing share',
          group_uuid: 'existing-share-id',
          permissions: ['order', 'read']
        }
      ]);
    mockApi.onGet(`${RBAC_API_BASE}/groups/`).replyOnce(200, {
      data: [
        {
          uuid: 'existing-share-id',
          name: 'Existing share'
        },
        {
          uuid: 'new-share-id',
          name: 'New share'
        }
      ]
    });
    jest.runAllTimers();
    await act(async () => {
      wrapper
        .find('li#share-portfolio-action a')
        .simulate('click', { button: 0 });
    });

    await act(async () => {
      wrapper.update();
      jest.runAllTimers();
    });
    /**
     * cear the existing share and select the new share
     */
    const clearButton = wrapper.find(
      '.ddorg__pf4-component-mapper__select__indicators .pf-c-button.pf-m-plain'
    );
    clearButton.simulate('click');
    clearButton.simulate('mousedown');
    wrapper.update();
    wrapper
      .find('div.ddorg__pf4-component-mapper__select__control')
      .first()
      .simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });

    expect(
      wrapper.find('div.ddorg__pf4-component-mapper__select__option')
    ).toHaveLength(2);

    wrapper.update();
    wrapper
      .find('div.ddorg__pf4-component-mapper__select__option')
      .last()
      .simulate('click');

    wrapper
      .find('div.ddorg__pf4-component-mapper__select__control')
      .at(1)
      .simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
    wrapper
      .find('div.ddorg__pf4-component-mapper__select__option')
      .last()
      .simulate('click');

    /**
     * mock share/ushare calls
     */
    mockApi
      .onPost(`${CATALOG_API_BASE}/portfolios/${initialPortfolio.id}/share`)
      .replyOnce((req) => {
        expect(JSON.parse(req.data)).toEqual({
          permissions: ['order', 'read'],
          group_uuids: ['new-share-id']
        });
        return [200];
      });
    mockApi
      .onPost(`${CATALOG_API_BASE}/portfolios/${initialPortfolio.id}/unshare`)
      .replyOnce((req) => {
        expect(JSON.parse(req.data)).toEqual({
          permissions: ['order', 'read'],
          group_uuids: ['existing-share-id']
        });
        return [200];
      });
    await act(async () => {
      wrapper.find('form').simulate('submit');
    });
    wrapper.update();
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolios');
    expect(wrapper.find(Portfolios)).toHaveLength(1);
    expect(wrapper.find(StyledGalleryItem)).toHaveLength(1);
    done();
  });

  it('should create, edit, copy, delete and undo delete of portfolio in app', async () => {
    const initialPortfolio = {
      id: '123',
      name: 'New portfolio',
      created_at: new Date().toString(),
      metadata: {
        user_capabilities: {
          create: true,
          copy: true,
          show: true,
          update: true,
          destroy: true,
          share: true,
          unshare: true
        }
      }
    };
    mockApi.onGet(`${RBAC_API_BASE}/groups/`).replyOnce(200, {
      data: [
        {
          uuid: 'existing-share-id',
          name: 'Existing share'
        },
        {
          uuid: 'new-share-id',
          name: 'New share'
        }
      ]
    });
    jest
      .spyOn(PortfolioHelper, 'fetchPortfolioByName')
      .mockImplementation(() => new Promise((res) => res({ data: [] })));
    jest.spyOn(global.insights.chrome, 'getUserPermissions').mockImplementation(
      () =>
        new Promise((res) =>
          res([
            {
              permission: 'catalog:portfolios:create'
            }
          ])
        )
    );
    let wrapper;
    /**
     * Mock initial open API request and souce types request
     */
    mockApi.onGet(`${CATALOG_API_BASE}/openapi.json`).replyOnce(200, {
      components: { schemas: {} }
    });
    mockApi
      .onGet(`${SOURCES_API_BASE}/source_types`)
      .replyOnce(200, { data: [] });
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/workflows?filter[name][contains]=&filter[id][]=111`
      )
      .replyOnce(200, { data: [] });

    /**
     * Mock initial portfolios request after user lands on portfolios page
     * Response is empty
     */
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, {
        data: [],
        meta: { count: 0, limit: 50, offset: 0 }
      });
    const store = testStore();
    await act(async () => {
      wrapper = mount(
        <Provider store={store}>
          <MemoryRouter initialEntries={['/']}>
            <App />
          </MemoryRouter>
        </Provider>
      );
    });
    /**
     * One update to wait for chuk load
     */
    await act(async () => {
      wrapper.update();
    });
    /**
     * One update to update the component state
     */
    await act(async () => {
      wrapper.update();
    });
    expect(wrapper.find(Portfolios)).toHaveLength(1);
    expect(wrapper.find(ContentGalleryEmptyState)).toHaveLength(1);
    expect(wrapper.find('a#create-portfolio')).toHaveLength(1);
    /**
     * Click on create portfolio link
     * Modal with create portfolio form should appear
     */
    wrapper.find('a#create-portfolio').simulate('click', { button: 0 });
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolios/add-portfolio');
    expect(wrapper.find(AddPortfolioModal)).toHaveLength(1);
    /**
     * change the portfolio name, wait for name async validation and submit the form
     */
    const addPortfolioNameInput = wrapper.find('input#name');
    /**
     * Mock portfolio async name validation
     */
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, {
        data: [initialPortfolio],
        meta: {
          limit: 50,
          offset: 0,
          count: 1
        }
      });
    /**
     * Mock portfolio endpoint call
     */
    mockApi
      .onGet(`${CATALOG_API_BASE}/portfolios/${initialPortfolio.id}`)
      .replyOnce(200, initialPortfolio);

    addPortfolioNameInput.getDOMNode().value = initialPortfolio.name;
    await act(async () => {
      addPortfolioNameInput.simulate('change');
    });
    jest.runAllTimers();
    /**
     * mock create portfolio endpoint
     */
    mockApi.onPost(`${CATALOG_API_BASE}/portfolios`).replyOnce(() => {
      return [200, initialPortfolio];
    });
    await act(async () => {
      wrapper.find('form').simulate('submit');
    });
    wrapper.update();
    await act(async () => {
      wrapper.update();
    });

    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios/${initialPortfolio.id}/portfolio_items?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, { data: [], meta: { limit: 50, offset: 0, count: 0 } });

    /**
     * async data loading update
     */
    await act(async () => {
      wrapper.update();
    });

    /**
     * The app should now be displaying the Portfolio component
     * It should have no portfolio items
     */
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolio');
    expect(
      wrapper.find(MemoryRouter).instance().history.location.search
    ).toEqual('?portfolio=123');
    /**
     * async data loading update
     */
    await act(async () => {
      wrapper.update();
    });
    /**
     * JS chunk loading
     */
    wrapper.update();
    expect(wrapper.find(Portfolio)).toHaveLength(1);
    expect(wrapper.find(PortfolioEmptyState)).toHaveLength(1);
    /**
     * open the edit modal and change the portfolio description
     */
    wrapper.find('button#toggle-portfolio-actions').simulate('click');
    wrapper
      .find('li#edit-portfolio')
      .find('a')
      .simulate('click', { button: 0 });
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolio/edit-portfolio');

    initialPortfolio.description = 'edited portfolio description' // eslint-disable-line
    wrapper.update();
    /**
     * Mock patch endpoint
     */
    mockApi
      .onPatch(`${CATALOG_API_BASE}/portfolios/123`)
      .replyOnce(200, initialPortfolio);
    const editPortfolioDescriptionInput = wrapper.find('textarea#description');
    editPortfolioDescriptionInput.getDOMNode().value =
      'edited portfolio description';
    await act(async () => {
      editPortfolioDescriptionInput.simulate('change');
    });
    jest.runAllTimers();
    /**
     * Mock new portfolios request
     */
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, {
        data: [initialPortfolio],
        meta: {
          limit: 50,
          offset: 0,
          count: 1
        }
      });
    mockApi
      .onGet(`${CATALOG_API_BASE}/portfolios/${initialPortfolio.id}`)
      .replyOnce(200, initialPortfolio);
    await act(async () => {
      wrapper.find('form').simulate('submit');
    });

    /**
     * portfolio should have updated description in toolbar
     */
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolio');
    /**
     * After portfolio update
     */
    wrapper.update();
    expect(
      toJson(wrapper.find('p.top-toolbar-title-description'))
    ).toMatchSnapshot();

    /**
     * Portfolio should be copied and app should be redirected to new portfolio
     */
    const copiedPortfolio = {
      ...initialPortfolio,
      id: '1234',
      name: `Copy of ${initialPortfolio.name}`
    };
    wrapper.find('button#toggle-portfolio-actions').simulate('click');
    /**
     * Mock copy portfolio requests
     */
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, {
        data: [initialPortfolio, copiedPortfolio],
        meta: { count: 2, limit: 50, offset: 0 }
      });
    mockApi
      .onPost(`${CATALOG_API_BASE}/portfolios/123/copy`)
      .replyOnce(200, copiedPortfolio);
    await act(async () => {
      wrapper.find('li#copy-portfolio').simulate('click');
    });
    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolio');
    expect(
      wrapper.find(MemoryRouter).instance().history.location.search
    ).toEqual('?portfolio=1234');

    /**
     * Navigate back to portfolios list via breadcrumbs
     * There should be two portfolios now
     */
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, {
        data: [initialPortfolio, copiedPortfolio],
        meta: { count: 2, limit: 50, offset: 0 }
      });
    await act(async () => {
      wrapper.find('a.pf-c-breadcrumb__item').simulate('click', { button: 0 });
    });
    /**
     * loader update
     */
    wrapper.update();
    expect(wrapper.find(StyledGalleryItem)).toHaveLength(2);
    /**
     * delete the first portfolio
     */
    wrapper
      .find(`div#portfolio-${initialPortfolio.id}-dropdown button`)
      .simulate('click');
    wrapper
      .find('li#remove-portfolio-action a')
      .simulate('click', { button: 0 });

    expect(
      wrapper.find(MemoryRouter).instance().history.location.pathname
    ).toEqual('/portfolios/remove');
    expect(
      wrapper.find(MemoryRouter).instance().history.location.search
    ).toEqual('?portfolio=123');

    expect(wrapper.find(RemovePortfolioModal)).toHaveLength(1);
    /**
     * Mock delete api requests
     */
    mockApi
      .onDelete(`${CATALOG_API_BASE}/portfolios/${initialPortfolio.id}`)
      .replyOnce(200);
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, {
        data: [copiedPortfolio],
        meta: { count: 1, limit: 50, offset: 0 }
      });
    await act(async () => {
      wrapper.find('button#confirm-delete-portfolio').simulate('click');
    });
    wrapper.update();
    expect(wrapper.find(StyledGalleryItem)).toHaveLength(2);
    /**
     * Undo delete of portfolio
     */
    expect(wrapper.find('a#undo-delete-portfolio-123')).toHaveLength(1);
    /**
     * Mock undo delete endpoints
     */
    mockApi
      .onPost(`${CATALOG_API_BASE}/portfolios/${initialPortfolio.id}/undelete`)
      .replyOnce(200, initialPortfolio);
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolios?filter[name][contains_i]=&limit=50&offset=0`
      )
      .replyOnce(200, {
        data: [initialPortfolio, copiedPortfolio],
        meta: { count: 2, limit: 50, offset: 0 }
      });
    await act(async () => {
      wrapper.find('a#undo-delete-portfolio-123').simulate('click');
    });
    expect(wrapper.find(StyledGalleryItem)).toHaveLength(2);
  });
});
