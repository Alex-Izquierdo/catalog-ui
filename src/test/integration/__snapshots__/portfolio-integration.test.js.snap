// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test for portfolio entity should create, edit, copy, delete and undo delete of portfolio in app 1`] = `
<p
  className="top-toolbar-title-description"
  data-pf-content={true}
>
  edited portfolio description
</p>
`;
